runtime/src/accounts_db.rs

```diff
@@ -2259,28 +2289,279 @@ impl AccountsDb {
+                // fetched from storage. This means a race occurred between this function and clean
+                // accounts/purge_slots
+                let message = format!(
+                    "do_load() failed to get key: {} from storage, latest attempt was for \
+                     slot: {}, storage_entry: {} is_root_fixed: {}",
+                    pubkey, slot, store_id, is_root_fixed,
+                );
+                datapoint_warn!("accounts_db-do_load_warn", ("warn", message, String));
+                return None;
+            }
+
+            // Because reading from the cache/storage failed, retry from the index read
+            let (new_slot, new_store_id, new_offset) =
+                self.read_index_for_accessor(ancestors, pubkey, max_root)?;
+            if new_slot == slot && new_store_id == store_id {
+                // Considering that we're failed to get acccessor above and further that
+                // the index still returned the same (slot, store_id) tuple, offset must be same
+                // too.
+                assert!(new_offset == offset);
+
+                // If the entry was missing from the cache, that means it must have been flushed,
+                // and the accounts index is always updated before cache flush, so store_id must
+                // not indicate being cached.
+                assert!(new_store_id != CACHE_VIRTUAL_STORAGE_ID);
+
+                // If this is not a cache entry, then this was a minor fork slot
+                // that had its storage entries cleaned up by purge_slots() but hasn't been
+                // cleaned yet. That means this must be rpc access and not replay/banking.
+                // But we can't guarantee `is_root_fixed' must be false, because
+                // the dangling bank at the tip of minor fork might not be started to freeze
+                // (!freeze_started =~> is_root_fixed)
+
+                // Everything being said, let's return None as it's an error condition after all.
+                return None;
+            }
+
+            slot = new_slot;
+            store_id = new_store_id;
+            offset = new_offset;
+        }
     }
 
     fn do_load(
         &self,
         ancestors: &Ancestors,
         pubkey: &Pubkey,
         max_root: Option<Slot>,
+        is_root_fixed: bool,
     ) -> Option<(AccountSharedData, Slot)> {
-        let (slot, store_id, offset) = {
-            let (lock, index) = self.accounts_index.get(pubkey, Some(ancestors), max_root)?;
-            let slot_list = lock.slot_list();
-            let (
-                slot,
-                AccountInfo {
-                    store_id, offset, ..
-                },
-            ) = slot_list[index];
-            (slot, store_id, offset)
-            // `lock` released here
-        };
+        let (slot, store_id, offset) = self.read_index_for_accessor(ancestors, pubkey, max_root)?;
```

Comment:
I moved this initial index read out of the retry function, thanks to the read cache.

Reply:
Yup, thanks for taking care of that conflict!